sqlglot/optimizer/eliminate_subqueries.py

```diff
@@ -1,11 +1,20 @@
+from __future__ import annotations
+
 import itertools
+import typing as t
 
 from sqlglot import expressions as exp
 from sqlglot.helper import find_new_name
-from sqlglot.optimizer.scope import build_scope
+from sqlglot.optimizer.scope import Scope, build_scope
+
+if t.TYPE_CHECKING:
+    from sqlglot._typing import E
+
+    ExistingCTEsMapping = t.Dict[E, str]
+    TakenNameMaping = t.Dict[t.Any, t.Union[Scope, E]]
 
 
-def eliminate_subqueries(expression):
+def eliminate_subqueries(expression: exp.Expression) -> exp.Expression:
```

Comment:
I believe we want this to be `E -> E` so that the argument's type matches the returned one. For example, if you passed a `Query` to this and invoked `selects` on the output of the rule, mypy would complain because `selects` doesn't appear on the base `Expression` class.

Reply:
I think most of the pattern around the codebase does not use `E`, right? For example, `simplify` is such a case.

However, your point makes sense regardless afaict, will revert to `E`

Reply:
You can't just generalize my argument for those other rules. For example, `simplify` may look similar but it doesn't always preserve the input type in its output, so you can't use `E` there:

```python
>>> from sqlglot import parse_one
>>> from sqlglot.optimizer.simplify import simplify
>>>
>>> simplify(parse_one("1 + 2"))  # Gets an `Add`, returns a `Literal`.
Literal(this=3, is_string=False)
```

But I wouldn't be surprised if we're doing the same thing for other rules where the `E -> E` logic applies, so perhaps if you have cycles it's worth taking a look to refine them in a followup PR.